src/linear_algebra/clifford_algebra/basic.lean

```diff
@@ -64,23 +64,113 @@ end clifford_algebra
+| ⟨a⟩ ⟨b⟩ := ⟨a + b⟩
+@[irreducible] private def neg : clifford_algebra Q → clifford_algebra Q | ⟨a⟩ := ⟨-a⟩
+@[irreducible] private def mul : clifford_algebra Q → clifford_algebra Q → clifford_algebra Q
+| ⟨a⟩ ⟨b⟩ := ⟨a * b⟩
+@[irreducible] private def smul {S : Type*} [comm_semiring S] [algebra S (tensor_algebra R M)] :
+  S → clifford_algebra Q → clifford_algebra Q
+| r ⟨a⟩ := ⟨r • a⟩
+
+instance : has_zero (clifford_algebra Q) := ⟨zero Q⟩
+instance : has_one (clifford_algebra Q) := ⟨one Q⟩
+instance : has_add (clifford_algebra Q) := ⟨add Q⟩
+instance : has_mul (clifford_algebra Q) := ⟨mul Q⟩
+instance : has_neg (clifford_algebra Q) := ⟨neg Q⟩
+instance {S : Type*} [comm_semiring S] [algebra S (tensor_algebra R M)] :
+  has_smul S (clifford_algebra Q) := ⟨smul Q⟩
+
+instance : inhabited (clifford_algebra Q) := ⟨0⟩
+
+lemma zero_def : (0 : clifford_algebra Q) = ⟨0⟩ := by { show zero Q = _, rw zero }
+lemma one_def : (1 : clifford_algebra Q) = ⟨1⟩ := by { show one Q = _, rw one }
+
+lemma add_def {a b : ring_quot (clifford_algebra.rel Q)} :
+  (⟨a⟩ + ⟨b⟩ : clifford_algebra Q) = ⟨a + b⟩ := by { show add Q _ _ = _, rw add }
+
+lemma mul_def {a b : ring_quot (clifford_algebra.rel Q)} :
+  (⟨a⟩ * ⟨b⟩ : clifford_algebra Q) = ⟨a * b⟩ := by { show mul Q _ _ = _, rw mul }
+
+lemma neg_def {a : ring_quot (clifford_algebra.rel Q)} :
+  (-⟨a⟩ : clifford_algebra Q) = ⟨-a⟩ := by { show neg Q _ = _, rw neg }
+
+lemma smul_def {S : Type*} [comm_semiring S] [algebra S (tensor_algebra R M)]
+  {s : S} {a : ring_quot (clifford_algebra.rel Q)} :
+  (s • ⟨a⟩ : clifford_algebra Q) = ⟨s • a⟩ := by { show smul Q _ _ = _, rw smul }
+
+instance : ring (clifford_algebra Q) :=
+{ add           := (+),
+  mul           := (*),
+  zero          := 0,
+  one           := 1,
+  add_assoc     := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, simp [add_def, add_assoc] },
+  zero_add      := by { rintros ⟨⟩, simp [zero_def, add_def], },
+  add_zero      := by { rintros ⟨⟩, simp [zero_def, add_def] },
+  add_comm      := by { rintros ⟨⟩ ⟨⟩, simp [add_def, add_comm], },
+  mul_assoc     := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, simp [mul_def, mul_assoc] },
+  one_mul       := by { rintros ⟨⟩, simp [one_def, mul_def], },
+  mul_one       := by { rintros ⟨⟩, simp [one_def, mul_def] },
+  left_distrib  := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, simp [add_def, mul_def, left_distrib] },
+  right_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, simp [add_def, mul_def, right_distrib] },
+  neg           := has_neg.neg,
+  add_left_neg  := by { rintros ⟨⟩, simp [neg_def, add_def, zero_def] },
+  nsmul         := (•),
+  nsmul_zero'   := by { rintros ⟨⟩, simp [smul_def, zero_def] },
+  nsmul_succ'   := by { rintros n ⟨⟩, simp [smul_def, add_def, add_mul, add_comm] },
+  zsmul         := (•),
+  zsmul_zero'   := by { rintros ⟨⟩, simp [smul_def, zero_def] },
+  zsmul_succ'   := by { rintros n ⟨⟩, simp [smul_def, add_def, add_mul, add_comm] },
+  zsmul_neg'    := by { rintros n ⟨⟩, simp [smul_def, neg_def, add_mul] } }
```

Comment:
There should be a shorter proof here with `function.injective.ring`

Reply:
Indeed. In the end it's not shorter because one needs to add more boilerplate to define `nat_cast`, `int_cast` and `npow` before `function.injective.ring` is happy, but still it's more satisfactory.